blscurve/bls_batch_verifier.nim

```diff
@@ -280,29 +241,37 @@ checksAndStackTracesOff:
     return contexts[start].merge(contexts[mid])
 
 proc batchVerifyParallel*(
-       batcher: var BatchedBLSVerifierCache,
+       cache: var BatchedBLSVerifierCache,
+       input: openArray[SignatureSet],
        secureRandomBytes: array[32, byte]
      ): bool {.sideeffect.} =
   ## Multithreaded batch verification
-  ## Requires OpenMP 3.0 (GCC 4.4, 2008)
-  let numSets = batcher.sets.len.uint32
+  ## If multithreaded with -d:openmp requires OpenMP 3.0 (GCC 4.4, 2008)
+  ## This will verify all the inputs (PublicKey, message, Signature) triplets
+  ##  at once and return true if verification is successful.
+  ## If unsuccessful:
+  ## - The input was empty
+  ## - One or more of the inputs was invalid on aggregation
+  ## - One or more of the inputs had an invalid signature
+  ## If knowing which input was problematic is required, they must be checked one by one.
+  let numSets = input.len.uint32
   if numSets == 0:
     # Spec precondition
     return false
 
   let numBatches = min(numSets, omp_get_max_threads().uint32)
 
   # Stage 0: Accumulators - setLen for noinit of seq
-  batcher.batchContexts.setLen(numBatches.int)
-  batcher.updateResults.setLen(numBatches.int)
+  cache.batchContexts.setLen(numBatches.int)
```

Comment:
if contexts.len > numbatches.len, should we shrink it? it keep it?

Reply:
`setLen` will shrink it (or extend it and reallocate it if needed)

Reply:
yes, but we could potentially skip the shrinking

Reply:
In that case the reserved memory stays the same and the code does only `mySeq.len = numBatches`.

https://github.com/nim-lang/Nim/blob/0956a99/lib/system/seqs_v2.nim#L85 (ARC seqs but I think the normal seq are the same).

Reply:
sort of, it reruns the zero-init on regrowth